### DISCUSSION TITLE
Typescript Definitions LTS

### DISCUSSION THREAD

**Discussion by tannerlinsley** [TanStack/query#362] [General]
If you use TS and you love React Query, then you likely want TS type defs to be available as first-class citizens in this repo. 

I have decided that I would *LOVE* to have a better DX for TS users, as long as there are enough of you to keep them up to date with the library.

This comes with a few disclaimers:
- I am not stating that I intend to always keep them up to date myself.
- I will try to give as much heads-up about changes as they come, but I will not wait around for types to become stable before making releases.
- I will not be making any major or feature level semver bumps just for type changes. Semver will continue to follow normal spec from a non TS perspective and when types are fixed, I will release patch-level semver bumps.
-  When Typescript issues arise, the people who have indicated here that they are willing to help will be assigned to triage and handle said issues.

If you would like to help and can pledge support for helping write, maintain, and triage maintenance for the types under these conditions, raise your hand (leave a comment below that you are willing help) !

If you have other questions, ask them here.

**Comment by nksaraf** [comment]
Hi @tannerlinsley, would definitely love to help with this! And if more people are involved, we should be able to keep them up-to date.

**Reply by tannerlinsley** [reply]
Thank you! As soon as we have a few more people on board, we can start coordinating the PR.

**Reply by lednhatkhanh** [reply]
I also would love to contribute, I'm also using react-query and would love to see the happens!

**Comment by lishine** [comment]
I am very happy about your decision and would love to help.

**Reply by tannerlinsley** [reply]
Nice! I appreciate your support here. As soon as we have a few more people on board, we can start coordinating the PR.

**Comment by vankhoawin** [comment]
Popping in from https://github.com/tannerlinsley/react-query/issues/277, count me in :)

**Reply by tannerlinsley** [reply]
Thanks @vnguyen94 :) As soon as we have a few more people on board, we can start coordinating the PR.

**Comment by tannerlinsley** [comment]
Let's start by locating the most up-to-date types out there. Do any of you know where those might be? Definitely Typed? Are there any open PRs to it?

**Reply by vankhoawin** [reply]
not at the moment: https://github.com/DefinitelyTyped/DefinitelyTyped/pulls?q=is%3Apr+is%3Aopen+react-query

**Reply by tannerlinsley** [reply]
Okay, it looks like they are up to date then? How should we go about importing them into the repo?

I would like it if all of the type-related stuff was in a root directory called `types` or something

**Reply by mskelton** [reply]
Yeah, I think pulling the types from Definitely Typed would be a good starting point.  Additionally, it might be good to reach out to those who have maintained the types on Definitely Typed and see if they would be willing to help with the migration and continued maintenance.

**Reply by tannerlinsley** [reply]
Right. I've left a link to this discussion in a few other issues and discussions about typescript, so hopefully they see that and come back here.

**Reply by vankhoawin** [reply]
agreed, i see 3 things:
1) as mentioned, raise awareness in other issues/discussions about the migration
1) pull types into react-query
2) deprecate types in DefinitelyTyped which will let users know to remove the @types dependency in the future https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package

**Comment by tannerlinsley** [comment]
Okay, I've pulled in the types to this repo and set things up to the best of my limited typescript knowledge. `dtslint` is passing too.

I'm going to publish a new patch version with the types and a few other things and hopefully we start getting some feedback on it pretty quick. 

If all of you could check over the types, try them out, and make sure we're doing all the right things, that would be great.

Thanks!

**Comment by Igorbek** [comment]
Here are the authors who worked on the definitions in DT: @lukaszfiszer @jacehensley @matteofrana @igorbek @vnguyen94 @youngbob @armanio123 @bostondv @lindskogen

**Comment by vankhoawin** [comment]
@tannerlinsley is there not a Github PR template we can use? While you did mention that you might not necessarily be adding types as you go on, I think other contributors should update them along with their PRs. Once you link your PR here, I wouldn't mind adding a PR template with steps to update TS types/tests when necessary.
